lib/spack/llnl/util/filesystem.py

```diff
@@ -1712,66 +1715,123 @@ def find(root, files, recursive=True):
-        answer.extend(found_files[search_file])
+    Does not search in directories below the specified depth.
 
-    return answer
+    If ``globs`` is a list, files matching earlier entries are placed
+    in the return value before files matching later entries.
+    """
+    if not os.path.exists(root):
+        return []
 
+    if max_depth is None:
+        max_depth = sys.maxsize
 
-@system_path_filter
-def _find_non_recursive(root, search_files):
-    # The variable here is **on purpose** a defaultdict as os.list_dir
-    # can return files in any order (does not preserve stability)
-    found_files = collections.defaultdict(list)
+    if isinstance(globs, str):
+        globs = [globs]
+    # Apply normcase to regular expressions and to the filenames:
+    # this respects case-sensitivity semantics of different OSes
+    # (e.g. file search is typically case-insensitive on Windows)
+    regexes = [re.compile(fnmatch.translate(os.path.normcase(x))) for x in globs]
 
-    # Make the path absolute to have absolute path returned
+    if not os.path.isdir(root):
+        raise ValueError(f"{root} is not a directory")
+
+    # Note later calls to os.scandir etc. return abspaths if the
+    # input is absolute, see https://docs.python.org/3/library/os.html#os.DirEntry.path
     root = os.path.abspath(root)
 
-    for search_file in search_files:
-        matches = glob.glob(os.path.join(root, search_file))
-        matches = [os.path.join(root, x) for x in matches]
-        found_files[search_file].extend(matches)
+    found_files = collections.defaultdict(list)
+
+    def _follow_link(p):
+        last = p
+        while os.path.islink(last):
+            last = resolve_link_target_relative_to_the_link(last)
+        return last
+
+    def _dir_id(p):
+        stat_result = os.stat(p, follow_symlinks=True)
+        # Note: on windows, st_ino is the file index and st_dev
+        # is the volume serial number. See
+        # https://github.com/python/cpython/blob/3.9/Python/fileutils.c
+        return (stat_result.st_ino, stat_result.st_dev)
+
+    visited_dirs = set([_dir_id(root)])
+
+    # Each queue item stores the depth and path
+    # BFS traversal is important to ensure repeatable results given
+    # that we allow following symlinks to sibling dirs that are not
+    # as deep: we return the shortest path to a given file
+    dir_queue = collections.deque([(0, root)])
+    while dir_queue:
+        depth, next_dir = dir_queue.pop()
+        try:
+            dir_iter = os.scandir(next_dir)
+        except OSError:
+            # Most commonly, this would be a permissions issue, for
+            # example if we are scanning an external directory like /usr
+            continue
+
+        with dir_iter:
+            for dir_entry in dir_iter:
+                try:
+                    it_is_a_dir = dir_entry.is_dir(follow_symlinks=True)
+
+                    if it_is_a_dir:
+                        dir_id = _dir_id(dir_entry.path)
```

Comment:
I would just stick to two try-catch blocks so (my mental) static analysis doesn't have to work hard to understand that in all cases `dir_id` is not unbound in the second `if it_is_a_dir` branch.